cmd/kubeadm/app/phases/certs/certs.go

```diff
@@ -35,35 +35,88 @@ import (
 	glog.V(1).Infoln("creating PKI assets")
-	certActions := []func(cfg *kubeadmapi.InitConfiguration) error{
-		CreateCACertAndKeyFiles,
-		CreateAPIServerCertAndKeyFiles,
-		CreateAPIServerKubeletClientCertAndKeyFiles,
-		CreateServiceAccountKeyAndPublicKeyFiles,
-		CreateFrontProxyCACertAndKeyFiles,
-		CreateFrontProxyClientCertAndKeyFiles,
-	}
-	etcdCertActions := []func(cfg *kubeadmapi.InitConfiguration) error{
-		CreateEtcdCACertAndKeyFiles,
-		CreateEtcdServerCertAndKeyFiles,
-		CreateEtcdPeerCertAndKeyFiles,
-		CreateEtcdHealthcheckClientCertAndKeyFiles,
-		CreateAPIServerEtcdClientCertAndKeyFiles,
-	}
 
-	if cfg.Etcd.Local != nil {
-		certActions = append(certActions, etcdCertActions...)
+	// This structure cannot handle multilevel CA hierarchies.
+	// This isn't a problem right now, but may become one in the future.
+
+	certTree, err := makeCertTreeFromMap(CertMap)
+	if err != nil {
+		return err
 	}
 
-	for _, action := range certActions {
-		err := action(cfg)
-		if err != nil {
+	for ca, daughters := range certTree {
+		if err := CreateCertificateChain(ca, daughters, cfg); err != nil {
 			return err
 		}
 	}
 
 	fmt.Printf("[certificates] valid certificates and keys now exist in %q\n", cfg.CertificatesDir)
 
+	// Service accounts are not x509 certs, so handled separately
+	if err := CreateServiceAccountKeyAndPublicKeyFiles(cfg); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func makeCertTreeFromMap(certs map[string]*KubeadmCert) (map[*KubeadmCert][]*KubeadmCert, error) {
+	caMap := make(map[*KubeadmCert][]*KubeadmCert)
+
+	for _, cert := range certs {
+		if cert.CAName == "" {
+			if _, ok := caMap[cert]; !ok {
+				caMap[cert] = []*KubeadmCert{}
+			}
+		} else {
+			ca, ok := certs[cert.CAName]
+			if !ok {
+				return nil, fmt.Errorf("Certificate %q references unknown CA %q", cert.Name, cert.CAName)
+			}
+			caMap[ca] = append(caMap[ca], cert)
+		}
+	}
+
+	return caMap, nil
+}
+
+//CreateCertificateChain creates a CA, then signs all the daughter certificates using the created CA.
+func CreateCertificateChain(caCertConfig *KubeadmCert, daughterConfigs []*KubeadmCert, mc *kubeadmapi.InitConfiguration) error {
+	glog.V(1).Infof("creating new %s certificate authority", caCertConfig.Name)
+	caCert, caKey, err := NewCACertAndKey()
+	if err != nil {
+		return err
+	}
+
+	if err := writeCertificateAuthorithyFilesIfNotExist(mc.CertificatesDir, caCertConfig.BaseName, caCert, caKey); err != nil {
+		return fmt.Errorf("couldn't create certificate for %s: %v", caCertConfig.Name, err)
+	}
+
+	for _, daughterCert := range daughterConfigs {
+		if err = createDaughterCernificate(daughterCert, mc, caCert, caKey); err != nil {
+			return fmt.Errorf("could not create %s daughter cert: %v", caCertConfig.Name, err)
+		}
+	}
+
+	return nil
+}
+
+func createDaughterCernificate(daughter *KubeadmCert, mc *kubeadmapi.InitConfiguration, caCert *x509.Certificate, caKey *rsa.PrivateKey) error {
```

Comment:
Typo in the function name. Cernificate => Certificate

I'd prefer this to be a method on the `*KubeadmCert` type:

`func (cert *KubeadmCert) createDaughterCertificate( mc *kubeadmapi.InitConfiguration, caCert *x509.Certificate, caKey *rsa.PrivateKey) error {`

This consolidates cert logic to the  `KubeadmCert` type.

Reply:
:+1: